src/core/io/src/4C_io_input_spec_builders.hpp

```diff
@@ -0,0 +1,882 @@
+    }
+
+    /**
+     * Legacy dat-style printing. This format is intended to be human-readable and is not used
+     * for further processing.
+     */
+    void print(std::ostream& stream, const InputParameterContainer& container) const
+    {
+      if (!required()) stream << "[";
+      spec_->print(stream, container);
+      if (!required()) stream << "]";
+    }
+
+    [[nodiscard]] const std::string& name() const { return data_.name; }
+
+    [[nodiscard]] const std::string& description() const { return data_.description; }
+
+    [[nodiscard]] bool required() const { return data_.required; }
+
+    [[nodiscard]] bool has_default_value() const { return data_.has_default_value; }
+
+   private:
+    CommonData data_;
+
+    //! Pointer to type-erased implementation.
+    std::unique_ptr<Internal::InputSpecTypeErasedInterface> spec_;
+  };
+
+  /**
+   * Helper functions to create InputSpec objects.
+   */
+  namespace InputSpecBuilders
+  {
+    template <typename StoredTypeIn>
+    struct ScalarData
+    {
+      using StoredType = StoredTypeIn;
+      /**
+       * An optional description of the value.
+       */
+      std::string description{};
+
+      /**
+       * Whether the value is required or optional.
+       */
+      std::optional<bool> required{};
+
+      /**
+       * The default value of the parameter. If this optional fields is set, the parameter is
+       * optional. If it is not set, the parameter is required.
+       */
+      std::optional<StoredType> default_value{};
+    };
+
+    template <typename StoredTypeIn, typename ScalarTypeIn>
+    struct VectorData
+    {
+      using StoredType = StoredTypeIn;
+      using ScalarType = ScalarTypeIn;
+
+      /**
+       * A function that determines the size of the vector. This function is called with the
+       * already parsed content of the input line, which may be used to query the size as the
+       * value of another parameter.
+       */
+      using SizeCallback = std::function<int(const InputParameterContainer&)>;
+
+      /**
+       * An optional description of the value.
+       */
+      std::string description{};
+
+      /**
+       * Whether the value is required or optional.
+       */
+      std::optional<bool> required{};
+
+      /**
+       * The default value of the parameter. If this optional fields is set, the parameter is
+       * optional. If it is not set, the parameter is required.
+       */
+      std::optional<StoredType> default_value{};
+
+      /**
+       * The size of the data. This can either be a fixed size or a callback that determines the
+       * size based on the value of another parameter.
+       *
+       * @note We use `int` instead of `size_t` since this is what most users write. Template
+       *      argument deduction will work a lot better with `int` than with `size_t`.
+       */
+      std::variant<int, SizeCallback> size;
+    };
+
+    template <typename StoredTypeIn>
+    struct SelectionData
```

Comment:
`ScalarData` and `VectorData` is very clear by it's name. `SelectionData` might need some additional docu.

Reply:
Actually, I don't even need this struct any longer due to my latest changes. Will be removed.